scripts/shared/e2e.sh

```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env bash
+
+## Process command line flags ##
+
+source /usr/share/shflags/shflags
+DEFINE_string 'cluster_settings' '' "Settings file to customize cluster deployments"
+DEFINE_string 'focus' '.*' 'Ginkgo focus for the E2E tests'
+FLAGS "$@" || exit $?
+eval set -- "${FLAGS_ARGV}"
+
+focus="${FLAGS_focus}"
+cluster_settings="${FLAGS_cluster_settings}"
+
+set -em -o pipefail
+
+source ${SCRIPTS_DIR}/lib/debug_functions
+source ${SCRIPTS_DIR}/lib/utils
+
+# Always source the shared cluster settings, to set defaults in case something wasn't set in the provided settings
+source "${SCRIPTS_DIR}/lib/cluster_settings"
+[[ -z "${cluster_settings}" ]] || source ${cluster_settings}
+
+### Functions ###
+
+function deploy_env_once() {
+    if with_context cluster3 kubectl wait --for=condition=Ready pods -l app=submariner-engine -n "${SUBM_NS}" --timeout=3s > /dev/null 2>&1; then
+        echo "Submariner already deployed, skipping deployment..."
+        return
+    fi
+
+    make deploy
+}
+
+function generate_context_flags() {
+    for cluster in "${!cluster_subm[@]}"; do
+        if [[ "${cluster_subm[$cluster]}" = "true" ]]; then
+            printf " -dp-context $cluster"
+        fi
+    done
+}
+
+function test_with_e2e_tests {
+    cd ${DAPPER_SOURCE}/test/e2e
+
+    go test -v -args -ginkgo.v -ginkgo.randomizeAllSpecs \
+        -submariner-namespace $SUBM_NS $(generate_context_flags) \
+        -ginkgo.noColor -ginkgo.reportPassed \
+        -ginkgo.focus "\[${focus}\]" \
+        -ginkgo.reportFile ${DAPPER_OUTPUT}/e2e-junit.xml 2>&1 | \
+        tee ${DAPPER_OUTPUT}/e2e-tests.log
+}
+
+### Main ###
+
+declare_kubeconfig
+
+deploy_env_once
+test_with_e2e_tests
+
+cat << EOM
+Your 3 virtual clusters are deployed and working properly with your local submariner source code, and can be accessed with:
+
+export KUBECONFIG=\$(echo \$(git rev-parse --show-toplevel)/output/kubeconfigs/kind-config-cluster{1..3} | sed 's/ /:/g')
+
+$ kubectl config use-context cluster1 # or cluster2, cluster3..
+
+To clean evertyhing up, just run: make cleanup
+EOM
```

Comment:
I'm not sure of the general usefulness of having this in shipyard. It seems this would only be useful for submariner. Lighthouse has to do additional core-dns setup in its _e2e.sh_.  Admiral and shipyard tests (in progress) don't deploy submariner.  Perhaps it would be better to make more utility functions, eg `test_with_e2e_tests`. I already added utility to print the clusters message.

Reply:
Actually lighthouse runs E2E tests so it makes sense to have this shared. The additional setup will be done in an intermediary step.
As for Admiral and shipyard, we'll have to see exactly what we run in e2e, but at least in shipyard we do deploy to test the deployment itself is working so it's not like we're going to remove it. In Admiral I see that you are adding some deployment capability. For that I suggest we can use cluster_settings to skip deployments on certain clusters as it's already possible, instead of running your own deployment procedure. Alternatively you can define your own `deploy` target there and still use the E2E code here.

If we need to make adjustments we can always tweak what we need of course, but we should make the typical use case shared IMHO and not just some functions which are scattered.

Reply:
After reviewing the scripts at https://github.com/submariner-io/admiral/pull/49 it seems indeed admiral can benefit from this change (well if we make sure it doesn't try to find subm :)) and just employ it's own `deploy` target.

Reply:
Let's try to converge into a single set of e2e scripts, if we need to readjust later... we can figure out.